File: catboost/libs/model/model_export/resources/binarize_float_features_nan_mode_max.py
def binarize_float_features(model, binary_features, float_features):
    binary_feature_index = 0
    for i in range(len(model.float_feature_borders)):
        for border in model.float_feature_borders[i]:
            binary_features[binary_feature_index] += 1 if ((float_features[model.float_features_index[i]] > border) or math.isnan(float_features[model.float_features_index[i]])) else 0
        binary_feature_index += 1
    return binary_feature_index

File: catboost/libs/model/model_export/resources/binarize_float_features_nan_mode_max.cpp
static void BinarizeFloatFeatures(
    const std::vector<float>& floatFeatures,
    const struct CatboostModel& model,
    std::vector<unsigned char>& binaryFeatures,
    unsigned int& binFeatureIndex) {
    for (size_t i = 0; i < model.FloatFeatureBorders.size(); ++i) {
        if (!model.FloatFeatureBorders[i].empty()) {
            for (const float border : model.FloatFeatureBorders[i]) {
                binaryFeatures[binFeatureIndex] += (unsigned char) ((floatFeatures[i] > border) || std::isnan(floatFeatures[i]));
            }
            ++binFeatureIndex;
        }
    }
}
